bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ResultRowAdapter.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.hbase.adapters.read;
+
+import com.google.api.core.InternalApi;
+import com.google.cloud.bigtable.data.v2.models.RowAdapter;
+import com.google.cloud.bigtable.hbase.util.ByteStringer;
+import com.google.cloud.bigtable.hbase.util.TimestampConverter;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.client.Result;
+
+/**
+ * Adapter for {@link RowAdapter} that uses {@link Result} to represent logical rows.
+ *
+ * <p>For internal use only - public for technical reasons.
+ */
+@InternalApi("For internal usage only")
+public class ResultRowAdapter implements RowAdapter<Result> {
+
+  @Override
+  public RowBuilder<Result> createRowBuilder() {
+    return new ResultRowBuilder();
+  }
+
+  @Override
+  public boolean isScanMarkerRow(Result result) {
+    return result.isEmpty();
+  }
+
+  @Override
+  public ByteString getKey(Result result) {
+    return ByteStringer.wrap(result.getRow());
+  }
+
+  public static class ResultRowBuilder implements RowBuilder<Result> {
```

Comment:
I think this should use a custom subclass of Result. The subclass should store the original ByteString alongside the Cells. isMarkerRow and getRowKey can downcast to get the appropriate info

Reply:
Thanks a lot for the review and suggestions.
I am not sure where we will be using cells fields from subclass, So kept a boolean field for isMarkerRow(). like this:
```java
static class RowResult extends Result {
    private final ByteString rowKey;
    private final boolean isMarkerRow;

    RowResult(ByteString rowKey, List<Cell> cells) {
      super();
      this.rowKey = rowKey;
      this.isMarkerRow = cells == null || cells.isEmpty();

      // all except default ctor of Result are private, So instantiating cells through copyFrom()
      // because value(), size(), isEmpty() rawCells(), advance(). are directly using Result's cells field.
      this.copyFrom(Result.create(cells));
    }

    ByteString getKey() {
      return rowKey;
    }

    boolean isMarkerRow() {
      return isMarkerRow;
    }
  }
```

Please let me know you thoughts on this.